From 6e640ee48575b4f452c0392c5ca74a6668e617fc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Wed, 25 Jun 2025 10:38:07 +0200
Subject: [PATCH 883/954] fix(compiler-rt): force usage of __popcountdi2
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

NOT UPSTREAMABLE

Enabling LTO cause issue as somme function may be removed during
optimization while being needed. Add the use attribute to those function
as a workaround.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 1bd5b16ba5d88d8770081664c6ae861da8c34285 fix(compiler-rt): force usage of __popcountdi2]
Change-Id: I3e0e7a98f634284be02914de3c4303209515e4d9
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/469925
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 lib/compiler-rt/builtins/popcountdi2.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/compiler-rt/builtins/popcountdi2.c b/lib/compiler-rt/builtins/popcountdi2.c
index 20dd0b0239..e050e38566 100644
--- a/lib/compiler-rt/builtins/popcountdi2.c
+++ b/lib/compiler-rt/builtins/popcountdi2.c
@@ -14,7 +14,7 @@
 
 // Returns: count of 1 bits
 
-COMPILER_RT_ABI int __popcountdi2(di_int a) {
+COMPILER_RT_ABI __attribute__((used)) int __popcountdi2(di_int a) {
   du_int x2 = (du_int)a;
   x2 = x2 - ((x2 >> 1) & 0x5555555555555555uLL);
   // Every 2 bits holds the sum of every pair of bits (32)
-- 
2.53.0

